yt/utilities/lib/cyoctree.pyx

```diff
@@ -0,0 +1,590 @@
+                    temp.node_id = z
+
+                    temp.start = split_arr[split_id]
+                    temp.end = split_arr[split_id + 1]
+
+                    self.nodes.push_back(temp)
+                    z+=1
+                    split_id+=1
+
+    @property
+    def size_bytes(self):
+        return sizeof(Node) * self.nodes.size()
+
+    @property
+    def max_depth(self):
+        return self._max_depth
+
+    @property
+    def data_version(self):
+        return self._data_version
+
+    @property
+    def num_particles(self):
+        return self._num_particles
+
+    @property
+    def n_ref(self):
+        return self._n_ref
+
+    @property
+    def num_octs(self):
+        return self.nodes.size()
+
+    @property
+    def over_refine_factor(self):
+        return self._over_refine_factor
+
+    @property
+    def idx(self):
+        return np.asarray(self._idx)
+
+    @over_refine_factor.setter
+    def over_refine_factor(self, value):
+        self._over_refine_factor = value
+        self._num_cells = 2**(3 * value)
+        self._num_cells_per_dim = 2**value
+
+    @property
+    def density_factor(self):
+        return self._density_factor
+
+    @density_factor.setter
+    def density_factor(self, value):
+        self._density_factor = value
+        self._num_children = 2**(3 * value)
+        self._num_children_per_dim  = 2**value
+
+    @property
+    def cell_positions(self):
+        cdef int i, j, z, k, l, num_leaves
+
+        # find all the leafs
+        num_leaves = 0
+        for i in range(self.nodes.size()):
+            if self.nodes[i].leaf == 1:
+                num_leaves += 1
+
+        cdef np.float64_t[:, :] pos = np.zeros((num_leaves*self._num_cells, 3),
+                                                dtype=float)
+        cdef double leftx, lefty, leftz, rightx, righty, rightz
+        z = 0
+        for i in range(self.nodes.size()):
+            if self.nodes[i].leaf == 0:
+                continue
+
+            leftx = self.nodes[i].left_edge[0]
+            lefty = self.nodes[i].left_edge[1]
+            leftz = self.nodes[i].left_edge[2]
+            rightx = self.nodes[i].right_edge[0]
+            righty = self.nodes[i].right_edge[1]
+            rightz = self.nodes[i].right_edge[2]
+
+            self.nodes[i].leaf_id = z
+
+            # we have to generate cell locations
+            dx = (rightx - leftx) / self._num_cells_per_dim
+            dy = (righty - lefty) / self._num_cells_per_dim
+            dz = (rightz - leftz) / self._num_cells_per_dim
+
+            for j in range(self._num_cells_per_dim):
+                for k in range(self._num_cells_per_dim):
+                    for l in range(self._num_cells_per_dim):
+                        pos[z, 0] = leftx + (j + 0.5) * dx
+                        pos[z, 1] = lefty + (k + 0.5) * dy
+                        pos[z, 2] = leftz + (l + 0.5) * dz
+                        z+=1
+
+        return np.asarray(pos)
+
```

Comment:
I should move the deposition functions below when I add the rest of the deposition operations, which should simplify this file